### PR TITLE
Reduce logging and show no-op summary

### LINKED ISSUE #8792: Default verbosity should not report each project's noop restore.
# Details about Problem

When restoring a repo with a lot of projects (>500), NuGet logs an output message for every project, even if the project is up-to-date.  Instead, it should log a message for any actual action taken and then a summary indicating how many projects are up-to-date and took part in no-op.

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): Any

NuGet version (x.x.x.xxx): Any

dotnet.exe --version (if appropriate): Any

VS version (if appropriate): Any

OS version (i.e. win10 v1607 (14393.321)): Any

Worked before? If so, with which NuGet version: Any

## Detailed repro steps so we can see the same problem

1. Restore a large number of projects

... The output scrolls by for a few seconds even if everything is up-to-date.

### Verbose Logs

Please include verbose logs (NuGet.exe <COMMAND> -verbosity detailed | dotnet.exe <COMMAND> --verbosity diag | etc...)

### Sample Project

Very helpful if you can zip a project and paste into this issue!

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8792
Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

* Lower the verbosity of the message logged when a project is done restoring (since its logged for no-op and non no-op
* Raise the verbosity of the message logged when a project is actually restored
* Log a new message at the end of restore indicating how many projects were up-to-date

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:  Manual

### CODE REVIEW

**Comment by jeffkl** [comment]
Sample output:
```
Microsoft (R) Build Engine version 16.4.0 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restore completed in 3.73 sec for D:\MyRepo\src\tools\ToolA\ToolA.csproj.
  Restore completed in 3.73 sec for D:\MyRepo\src\tools\ToolA.UnitTests\ToolA.UnitTests.csproj.
  739 of 741 projects are up-to-date for restore.

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:08.66
```

Instead of getting 739 log messages, I get one for each project that was actually restored and a summary at the end confirming that the rest of the projects are up-to-date

**Review by nkolev92 — COMMENTED** [review]
I have a concern with this change, because NuGet provides no updates while the operation is happening. 
The 1 line per project experience provides clear progress. 
We can consider a summary, but I'm not sure that needs to change.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreSummary.cs [hunk 1]
```diff
@@ -74,13 +74,48 @@ public RestoreSummary(bool success)
             Errors = errors;
         }
 
-        public static void Log(ILogger logger, IEnumerable<RestoreSummary> restoreSummaries, bool logErrors = false)
+        public static void Log(ILogger logger, IReadOnlyList<RestoreSummary> restoreSummaries, bool logErrors = false)
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreSummary.cs:82]: This is a breaking API change. Overload?

**Comment by jeffkl** [comment]
@nkolev92 I added a new task that can log a localizable message and am calling it before restore so that users see a message while they're waiting for the MSBuild evaluations to finish.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreSummary.cs (hunk 1 above)
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreSummary.cs:82]: Yes good catch

**Review by dtivel — COMMENTED** [review]
Looks great, especially if it improves performance.  However, I'd like to see some tests verify the new strings and project counts.

**Comment by jeffkl** [comment]
@dtivel I have added unit tests

File: src/NuGet.Core/NuGet.Commands/Strings.resx [hunk 2]
```diff
@@ -230,6 +230,14 @@
   <data name="Log_InstalledSummary" xml:space="preserve">
     <value>Installed:</value>
   </data>
+  <data name="Log_AllProjectsUpToDate" xml:space="preserve">
+    <value>All projects are up-to-date</value>
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Commands/Strings.resx]: `All projects are up to date.`

Note the added period and the removal of hyphens.

File: src/NuGet.Core/NuGet.Commands/Strings.resx [hunk 3]
```diff
@@ -230,6 +230,14 @@
...
+  </data>
+  <data name="Log_ProjectUpToDateSummary" xml:space="preserve">
+    <value>{0} of {1} projects are up-to-date</value>
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Commands/Strings.resx]: `{0} of {1} projects are up to date.`

Note the added period and the removal of hyphens.

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Core/NuGet.Build.Tasks/NuGetMessageTask.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        [Required]
+        public string Name { get; set; }
+
+        public override bool Execute()
+        {
+            MessageImportance messageImportance = MessageImportance.Normal;
+
+            if (!string.IsNullOrWhiteSpace(Importance) && !Enum.TryParse(Importance, ignoreCase: true, out messageImportance))
+            {
+                messageImportance = MessageImportance.Normal;
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGetMessageTask.cs:32]: `messageImportance` is already set to `Normal` in line no. 29. I am not sure if this is the right enum value to assign.

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGetMessageTask.cs:32]: If `Enum.TryParse` fails, it sets the out variable to `0` which is `High`

https://github.com/microsoft/msbuild/blob/e70a3159d64f9ed6ec3b60253ef863fa883a99b1/src/Framework/BuildMessageEventArgs.cs#L15

**Review by AndyGerlicher — APPROVED** [review]
Looks great! I think people will really like reducing the excessive high-priority messages on restore!

**Comment by jeffkl** [comment]
Notes from cross team discussion:

- [x]  Try to log the same message if a project was out of date
- [x] Change the messages to indicate that projects were up-to-date in regards to restore only, not for build

**Comment by jeffkl** [comment]
@nkolev92 @rrelyea @zivkan Please take a look as this is ready for review.

**Comment by zivkan** [comment]
Looks good to me, but I won't approve until we create a release-5.5.x branch, and dev branch starts targeting VS 16.6

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs [hunk 5]
```diff
@@ -189,14 +189,14 @@ public async Task DotnetRestore_OneLinePerRestore()
...
-                Assert.True(1 == result.AllOutput.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries).Length, result.AllOutput);
+                Assert.True(2 == result.AllOutput.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries).Length, result.AllOutput);
 
                 // Act - make sure no-op does the same thing.
                 result = _msbuildFixture.RunDotnet(pathContext.SolutionRoot, $"restore proj.sln {$"--source \"{pathContext.PackageSource}\""}", ignoreExitCode: true);
 
                 // Assert
                 Assert.True(result.ExitCode == 0);
-                Assert.True(1 == result.AllOutput.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries).Length, result.AllOutput);
+                Assert.True(2 == result.AllOutput.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries).Length, result.AllOutput);
```
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetRestoreTests.cs:199]: Can you please add a test where more than 1 project noops and check that the number of lines is 2. (Do at least 3 projects, cause 2 is the old behavior.)

**Review by nkolev92 — COMMENTED** [review]
This fixes https://github.com/NuGet/Home/issues/6384 as well. 

@jeffkl Can you please take a second look at that issue, and ensure I am interpreting things correctly.

**Comment by jeffkl** [comment]
@nkolev92 I don't think [this change](https://github.com/NuGet/NuGet.Client/pull/3111/files#diff-077ba718d0cb16337a398450160253d8R270) will affect the output in VS since the logger seems to treat Minimal and Information as the same thing.

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreOperationLogger.cs#L496-L498

There is a minimal verbosity level in the enum but it is only used in a few places.  We'd need to update this logger to know about Minimal vs Information in order to fix https://github.com/NuGet/Home/issues/6384

**Comment by nkolev92** [comment]
I have removed your assignment from the issue.

**Review by zivkan — APPROVED** [review]
ok with me to merge now.

**Pull request merged**